packages/svelte/src/chat.svelte.ts

```diff
@@ -0,0 +1,337 @@
+import {
+  fillMessageParts,
+  generateId,
+  type UIMessage,
+  type UseChatOptions,
+  type JSONValue,
+  type ChatRequest,
+  extractMaxToolInvocationStep,
+  callChatApi,
+  shouldResubmitMessages,
+  type Message,
+  type CreateMessage,
+  type ChatRequestOptions,
+  prepareAttachmentsForRequest,
+  getMessageParts,
+  updateToolCallResult,
+  isAssistantMessageWithCompletedToolCalls,
+} from '@ai-sdk/ui-utils';
+import { isAbortError } from '@ai-sdk/provider-utils';
+import {
+  KeyedChatStore,
+  getChatContext,
+  hasChatContext,
+} from './chat-context.svelte.js';
+
+export type ChatOptions = Readonly<
+  Omit<UseChatOptions, 'keepLastMessageOnError'> & {
+    /**
+     * Maximum number of sequential LLM calls (steps), e.g. when you use tool calls.
+     * Must be at least 1.
+     * A maximum number is required to prevent infinite loops in the case of misconfigured tools.
+     * By default, it's set to 1, which means that only a single LLM call is made.
+     * @default 1
+     */
+    maxSteps?: number;
+  }
+>;
+
+export type { CreateMessage, Message, UIMessage };
+
+export class Chat {
+  readonly #options: ChatOptions = {};
+  readonly #api = $derived(this.#options.api ?? '/api/chat');
+  readonly #generateId = $derived(this.#options.generateId ?? generateId);
+  readonly #maxSteps = $derived(this.#options.maxSteps ?? 1);
+  readonly #streamProtocol = $derived(this.#options.streamProtocol ?? 'data');
+  readonly #keyedStore = $state<KeyedChatStore>()!;
```

Comment:
Does this need to be a `$state`?

```suggestion
  readonly #keyedStore: KeyedChatStore = null!
```

Reply:
Technically no but pragmatically yes -- if it's not a state then `#store` complains about using `#keyedStore` before it's assigned 😢

Reply:
@Rich-Harris this is another example of why we need to be able to declare `$state` and `$derived` in the constructor -- it's not possible to resolve use before assignment stuff like this without unneeded state declarations or TS ignores.